pkg/operator/staticpod/controller/prune/prune_controller.go

```diff
@@ -92,85 +85,61 @@ func getRevisionLimits(operatorSpec *operatorv1.StaticPodOperatorSpec) (int32, i
-
-	configMaps, err := c.configMapGetter.ConfigMaps(c.targetNamespace).List(ctx, metav1.ListOptions{})
-	if err != nil {
-		return []int{}, err
-	}
-	for _, configMap := range configMaps.Items {
-		if !strings.HasPrefix(configMap.Name, statusConfigMapName) {
-			continue
+// revisionsToKeep approximates the set of revisions to keep: spec.failedRevisionsLimit for failed revisions,
+// spec.succeededRevisionsLimit for succeed revisions (for all nodes). The approximation goes by:
+// - don't prune LatestAvailableRevision and the max(spec.failedRevisionLimit, spec.succeededRevisionLimit) - 1 revisions before it.
+// - don't prune a node's CurrentRevision and the spec.succeededRevisionLimit - 1 revisions before it.
+// - don't prune a node's TargetRevision and the spec.failedRevisionLimit - 1 revisions before it.
+// - don't prune a node's LastFailedRevision and the spec.failedRevisionLimit - 1 revisions before it.
+func (c *PruneController) revisionsToKeep(status *operatorv1.StaticPodOperatorStatus, failedLimit, succeededLimit int) sets.Int32 {
+	keep := sets.Int32{}
+	keep.Insert(int32RangeBelowOrEqual(status.LatestAvailableRevision, max(failedLimit, succeededLimit))...) // max because we don't know about failure or success
+	for _, ns := range status.NodeStatuses {
+		if ns.CurrentRevision != 0 {
+			keep.Insert(int32RangeBelowOrEqual(ns.CurrentRevision, succeededLimit)...)
 		}
-
-		if revision, ok := configMap.Data["revision"]; ok {
-			revisionNumber, err := strconv.Atoi(revision)
-			if err != nil {
-				return []int{}, err
-			}
-			switch configMap.Data["status"] {
-			case string(corev1.PodSucceeded):
-				succeededRevisions = append(succeededRevisions, revisionNumber)
-			case string(corev1.PodFailed):
-				failedRevisions = append(failedRevisions, revisionNumber)
-			case StatusAbandoned:
-				abandonedRevisions = append(abandonedRevisions, revisionNumber)
-			case StatusInProgress:
-				// we always protect inprogress
-				inProgressRevisions = append(inProgressRevisions, revisionNumber)
-
-			default:
-				// protect things you don't understand
-				unknownStatusRevisions = append(unknownStatusRevisions, revisionNumber)
-				recorder.Event("UnknownRevisionStatus", fmt.Sprintf("unknown status for revision %d: %v", revisionNumber, configMap.Data["status"]))
-			}
+		if ns.TargetRevision != 0 {
+			keep.Insert(int32RangeBelowOrEqual(ns.TargetRevision, failedLimit)...)
+		}
+		if ns.LastFailedRevision != 0 {
+			keep.Insert(int32RangeBelowOrEqual(ns.LastFailedRevision, failedLimit)...)
 		}
 	}
+	return keep
+}
 
-	// Return early if nothing to prune
-	if len(succeededRevisions)+len(failedRevisions)+len(abandonedRevisions) == 0 {
-		klog.V(2).Info("no revision IDs currently eligible to prune")
-		return []int{}, nil
+// int32Range returns range of int32 from upper-num+1 to upper. Num -1 means all up to upper.
+func int32RangeBelowOrEqual(upper int32, num int) []int32 {
```

Comment:
if `upper` is equal `1` (CurrentRevision) and the `num` is `5` (default limit) it will return negative numbers

Reply:
does it hurt? But let me cap to strickly positive.

Reply:
done